llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -1084,6 +1084,40 @@ void VPlanTransforms::simplifyRecipes(VPlan &Plan, Type &CanonicalIVTy) {
   }
 }
 
+static void convertToUniformRecipes(VPlan &Plan) {
+  auto TryToNarrow = [](VPBasicBlock *VPBB) {
+    for (VPRecipeBase &R : make_early_inc_range(reverse(*VPBB))) {
+      // Try to narrow wide and replicating recipes to uniform recipes, based on
+      // VPlan analysis.
+      auto *Def = dyn_cast<VPSingleDefRecipe>(&R);
+      if (!Def || !isa<VPReplicateRecipe, VPWidenRecipe>(Def) ||
+          !Def->getUnderlyingValue())
```

Comment:
```suggestion
```
can be asserted instead if desired - these recipes must have underlying values - in order to know what to replicate or widen.

Reply:
It could be asserted for `VPReplicateRecipe`, but not for `VPWidenRecipe` which does not require a underlying value.

Reply:
Hmm, if `VPWidenRecipe` uses the (optional) underlying value only for metadata and FMF, could it be replaced with `VPInstruction`?

Reply:
Yep we should be able to do that soon. I'll give it a try, need to see if we rely on the encoded facts that certain recipes are widened in various places.